### ISSUE TITLE
AlexNet/ ZFNet 코드 리뷰

### ISSUE THREAD

**Issue by hoiho77** [yskim0/pytorch-cifar10#1]
pytorch는 처음이라서 그냥 보는데 전체적으로 주석을 잘 달아주셔서 코드 가독성이 좋은 것 같아요!

## data_loader.py 

transform.Normalize()의 변수는 직접 계산하셔서 입력하신 건가요?(파이토치를 잘 몰라서,,)
transform에서 데이터를 resize할 때 alexnet은 입력값이 227이라고 알고 있는데 사이즈를 256? 224? 조정하신 건가요?? centercrop은 데이터를 자른 것인지!? (이렇게 전처리한 이유는 무엇인지 궁금하네요!!) 

cifar10 데이터셋이 제 컴에서 잘 안돌아가서 일부만 활용하느라 데이터셋 load를 그냥 너무 단순하게 한 것 같아서 저도 shuffle이나 augmentation 등 다양한 옵션을 고민해봐야겠네요

## alexnet, zfnet
classifier과 feature extraction을 따로 구현해서 이 tensorflow랑 달라서 신기하네요.
forward() 도 신기하구요! 근데 이 함수가 호출이 안 되어도 되는건지?(forward함수 호출하는걸 못찾겠어서.. 그냥 실행이 되는 건가요?!! 제가 이해를 잘못한 거 일 수도 있어요!ㅜㅜ ) 

## train.py

logging.info() 저도 넣어야겠다는 생각이 드네요ㅎㅎ
tensorflow-gpu로 할 때는 gpu가 그냥 구동이되는 것 같은데 
pytorch라서 다른 건지 아니면 저렇게 정할 수 있는 옵션이 저도 되는지 한번 시도해봐야겠어요! tensorflow는 그냥 모델 돌리기엔 편한데 pytorch로 보면 확실히 여러 단계들을 세분화해서 볼 수 있다는 생각이 드네요!

## utils.py
각각 모델 별로 optimizer을 다르게 로드하게 해두셔서 좋다는 생각을 했습니다!!!
alexnet은 adam인데 제가 sgd로 잘못했다는 것도 발견했네요.
save_checkpoints( )랑 best 모델 저장하는 것도 저도 적용해야겠어요!!

----

고생하셨습니다 👐 여러모로 배울 부분과 제게 적용해야할 부분이 많네요ㅎㅎ
제가 아직 공부할 게 많아서..ㅎㅎ 혹시 질문한 부분 중에 답주실 수 있으신 부분은 시간 나실 때 언제든 답 주셔도 좋아요ㅎㅎ 앞으로도 화이팅!!

**Comment by yskim0** [comment]
우선 자세한 코드리뷰 감사합니다! 😄

## data_loader.py

> transform.Normalize()의 변수는 직접 계산하셔서 입력하신 건가요?(파이토치를 잘 몰라서,,)
transform에서 데이터를 resize할 때 alexnet은 입력값이 227이라고 알고 있는데 사이즈를 256? 224? 조정하신 건가요?? centercrop은 데이터를 자른 것인지!? (이렇게 전처리한 이유는 무엇인지 궁금하네요!!)
cifar10 데이터셋이 제 컴에서 잘 안돌아가서 일부만 활용하느라 데이터셋 load를 그냥 너무 단순하게 한 것 같아서 저도 shuffle이나 augmentation 등 다양한 옵션을 고민해봐야겠네요

- Normalize 안에 있는 숫자들은 ,, CIFAR-10이 워낙 유명한 데이터셋이다 보니 그냥 검색하니까 바로 나오더라구요. 그대로 집어 넣었습니다 ㅋㅋ 텐서플로우에서는 이를 자동으로 계산해주는 메서드가 있었죠? 파이토치도 있는지 봐야겠네요..

- 256으로 키운다음에 `CenterCrop`으로 224로 줄였습니다 근데 input 224아닌가요? (아래 사진은 구글링해서 가져왔습니다. 사실 알렉스넷은 논문을 보고 구현한 건 아니라 틀렸을 수도 있겠네요 ㅜ)
![](https://www.researchgate.net/profile/Jaime_Gallego2/publication/318168077/figure/fig1/AS:578190894927872@1514862859810/AlexNet-CNN-architecture-layers.png)

아무튼 한번에 224로 하지 않고 늘렸다가 crop한 이유는 ,,, 좋은 코드들 참고하니까 거의 그렇게 하더라구요. 검색해보니 근거가 없는 짓은 아니네요!
>> it seems to me that whomever created this transform function is expecting non-uniform input image. **the Resize functions is there only to ensure uniformity, and center crop reduce background noise.** [출처](https://discuss.pytorch.org/t/transforms-resize-vs-centercrop/86588)

보통 이미지 데이터셋들은 uniform하지 않기 때문에 그를 위해 이런 transform 과정을 거치는 듯 합니다.

넵 이미지는 모델에 들어가기 전에 여러 transform을 생각할 수 있고, 이에 따라 성능이 바뀌기도 하니까 고민하는 건 매우 좋은 일이죠!ㅎㅎ

**Comment by yskim0** [comment]
## alexnet, zfnet
> classifier과 feature extraction을 따로 구현해서 이 tensorflow랑 달라서 신기하네요.
forward() 도 신기하구요! 근데 이 함수가 호출이 안 되어도 되는건지?(forward함수 호출하는걸 못찾겠어서.. 그냥 실행이 되는 건가요?!! 제가 이해를 잘못한 거 일 수도 있어요!ㅜㅜ )

넵 맞습니다.
파이토치의 모델 클래스들은 모두 `nn.Module` 클래스를 상속하며, `__init__, forward` 메서드가 반드시 존재해야 합니다.
>> forward() 는 모델이 학습데이터를 입력받아서 forward propagation을 진행시키는 함수이고, 반드시 forward 라는 이름의 함수이어야 합니다.

>> nn.Module you will see that __call__ internally invokes forward

내재적으로? 자동적으로 `forward`가 호출됩니다!

## train.py
> logging.info() 저도 넣어야겠다는 생각이 드네요ㅎㅎ
tensorflow-gpu로 할 때는 gpu가 그냥 구동이되는 것 같은데
pytorch라서 다른 건지 아니면 저렇게 정할 수 있는 옵션이 저도 되는지 한번 시도해봐야겠어요! tensorflow는 그냥 모델 돌리기엔 편한데 pytorch로 보면 확실히 여러 단계들을 세분화해서 볼 수 있다는 생각이 드네요!

`train.py`에 사용자가 gpu를 사용한다고 했는지 안했는지 argument로 확인한 다음에 `.cuda()`로 gpu계산을 해주게 됩니다!
`.cuda()`나 device에 GPU를 선언하고, `.to(device)`를 하지 않으면 토치에서는 GPU로 동작하지 않아요!

## utils.py
> 각각 모델 별로 optimizer을 다르게 로드하게 해두셔서 좋다는 생각을 했습니다!!!
alexnet은 adam인데 제가 sgd로 잘못했다는 것도 발견했네요.
save_checkpoints( )랑 best 모델 저장하는 것도 저도 적용해야겠어요!!

감사합니다 ㅎㅎ ZFNet이 옵티마이저를 다른 걸 써서 저 함수를 새로 만들었네요 ㅠㅠ (사실 귀찮았어요ㅋㅋㅋ)
체크포인트 틈틈히 저장하는 거 좋습니다! 중간에 갑자기 꺼질 수도 있고, 오류 날 수도 있으니까요 ㅎㅎ

**Comment by yskim0** [comment]
일단 close하겠습니다.
추가 질문은 comment 달아주세요!
